### PR TITLE
Add geocoding to the hc client

### LINKED ISSUE #1221: Add Geocoding to HC Client
Currently, if one wants to use the GraphHopper routing and geocoding API, a user either has to only the swagger client (no points encoded etc.) or has to mix the hc client with the swagger client. I think it would be a nice addition to the hc client to be able to use geocoding as well.

### PR DESCRIPTION
**Pull request by boldtrn**
This PR fixes #1221.

I reused the model classes that we created in the geocoding converter to easily convert json received via the API to Java objects. Probably the effort of hosting them in an independent repository and add a dependency from gh and the geocoding converter is not worth it. Also I changed one or two small details like the name of the classes and removed a constructor that used AbstractAddress, which we don't need here.

### CODE REVIEW

File: client-hc/src/main/java/com/graphhopper/api/GraphHopperGeocoding.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+        if (request.isReverse()) {
+            url += "reverse=true";
+            url += "&point=" + request.getLat() + "," + request.getLon();
+        } else {
+            url += "reverse=false";
+            url += "&q=" + request.getQuery();
+            if (!Double.isNaN(request.getLat()))
+                url += "&point=" + request.getLat() + "," + request.getLon();
+        }
```
Comment (karussell) [inline client-hc/src/main/java/com/graphhopper/api/GraphHopperGeocoding.java]: I would help here the user a bit: if reverse than throw an exception if the lat or lon is invalid. And if non-reverse throw exc if query is empty. Also adding the point can be moved outside of the if+else IMO to have just one line

File: client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        return reverse;
+    }
+
+    public double getLat() {
+        return lat;
+    }
+
+    public double getLon() {
+        return lon;
+    }
```
Comment (karussell) [inline client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Not sure if important: we could reuse the GHPoint class here too

Comment (boldtrn) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Ok, sure, I can do this. Not sure if it's a big benefit. Should I change it?

File: client-hc/src/test/java/com/graphhopper/api/GraphHopperGeocodingIT.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.net.SocketTimeoutException;
+
+import static org.junit.Assert.*;
+
+/**
+ * @author Robin Boldt
+ */
+public class GraphHopperGeocodingIT {
+
+    public static final String KEY = "614b8305-b4db-48c9-bf4a-40de90919939";
```
Comment (karussell) [inline client-hc/src/test/java/com/graphhopper/api/GraphHopperGeocodingIT.java]: We should avoid duplicating this info and reuse GraphHopperWebIT.KEY

File: client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java (hunk 2 above)
Comment (karussell) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Is it necessary for serialization?

**Comment by karussell** [comment]
Also this name would need a change https://github.com/graphhopper/graphhopper/blob/master/client-hc/pom.xml#L28

File: client-hc/src/test/java/com/graphhopper/api/GraphHopperGeocodingIT.java (hunk 3 above)
Comment (boldtrn) [reply client-hc/src/test/java/com/graphhopper/api/GraphHopperGeocodingIT.java]: Thanks, I missed that :+1:

File: client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java (hunk 2 above)
Comment (boldtrn) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: No, we only serialize the response from the server.

File: client-hc/src/main/java/com/graphhopper/api/GraphHopperGeocoding.java (hunk 1 above)
Comment (boldtrn) [reply client-hc/src/main/java/com/graphhopper/api/GraphHopperGeocoding.java]: Thanks, good idea!

File: client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java (hunk 2 above)
Comment (boldtrn) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: I haven't changed that so far, let me know if I should

**Comment by boldtrn** [comment]
I updated this PR also adding some more documentation to make it easier for users to use methods that are publicly available.

File: client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java (hunk 2 above)
Comment (karussell) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Ok, yes, please change it to use GHPoint

Comment (karussell) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: This way we can just check against null instead of the NaN checks

Comment (karussell) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Or just point.isValid

Comment (boldtrn) [reply client-hc/src/main/java/com/graphhopper/api/model/GHGeocodingRequest.java]: Good idea, I updated the code :)

**Comment by karussell** [comment]
Thanks!

**Pull request merged**